In [81]:
import pandas as pd
import numpy as np

In [82]:
data = {
    'order_id':[1001,1002,1003,1004,1005,1006,1007],
    'order_date':['2023-01-15','2023-01-18','2023-02-10','2023-02-20','2023-03-05','2023-03-15','2023-03-25'],
    'customer_name':['Alice','Bob','Charlie','Alice','Eve','Frank','Grace'],
    'Region':['North','South','East','North','South','East','West'],
    'category':['Furniture','Technology','Office supplier','Furniture','Technology','Office supplier','Furniture'],
    'Product_name':['chair','laptop','paper','table','monitor','stapler','sofa'],
    'sales':[250,1200,80,450,900,30,670],
    'quantity':[2,1,10,1,2,20,1],
    'Revenue':[500.0,1200.0,800.0,450.0,np.nan,600.0,670.0]
}
df = pd.DataFrame(data)

In [83]:
data

{'order_id': [1001, 1002, 1003, 1004, 1005, 1006, 1007],
 'order_date': ['2023-01-15',
  '2023-01-18',
  '2023-02-10',
  '2023-02-20',
  '2023-03-05',
  '2023-03-15',
  '2023-03-25'],
 'customer_name': ['Alice',
  'Bob',
  'Charlie',
  'Alice',
  'Eve',
  'Frank',
  'Grace'],
 'Region': ['North', 'South', 'East', 'North', 'South', 'East', 'West'],
 'category': ['Furniture',
  'Technology',
  'Office supplier',
  'Furniture',
  'Technology',
  'Office supplier',
  'Furniture'],
 'Product_name': ['chair',
  'laptop',
  'paper',
  'table',
  'monitor',
  'stapler',
  'sofa'],
 'sales': [250, 1200, 80, 450, 900, 30, 670],
 'quantity': [2, 1, 10, 1, 2, 20, 1],
 'Revenue': [500.0, 1200.0, 800.0, 450.0, nan, 600.0, 670.0]}

1. How does apply() differ from map() and applymap()? When would you use each?
- apply() function can used for complex operations and function applies along the axis of a Dataframe whereas map() function applies cuntion to a series element wise and applymap() is also applies element wise to a series.
---------------------------------------------------------------------------------------------------------------------------------------------------------

2. Interpret the describe() method table for this dataset

In [86]:
df.describe()

order_id        sales   quantity      Revenue
count     7.000000     7.000000   7.000000     6.000000
mean   1004.000000   511.428571   5.285714   703.333333
std       2.160247   435.944514   7.250616   273.252020
min    1001.000000    30.000000   1.000000   450.000000
25%    1002.500000   165.000000   1.000000   525.000000
50%    1004.000000   450.000000   2.000000   635.000000
75%    1005.500000   785.000000   6.000000   767.500000
max    1007.000000  1200.000000  20.000000  1200.000000

- order_id
    - there are total 7 orders have been placed.
    - average order id is 1004.
    - maximum order id is 1007 and minimum id is 1001
- sales
    - Total sales count is 7.
    - The average sales are present around 511.428.
    - With standard deviation of 435.944 shows variation in sales
    - Minimum sales present with 30 count while maximum sales have placed with count of 1200.
- quantity
    - Total 7 quantity count is present in the dataset.
    - Average quantity of products has purchased by customers around 5.2857
    - 7.25 standard deviation shows customers purchased more products.
---------------------------------------------------------------------------------------------------------------------------------------------------------

3. Group the sales by 'Region' and calculate the average 'Sales'

In [89]:
# average of sales
avg = df['sales'].mean()
print('Average of sales = ',avg)

Average of sales =  511.42857142857144


In [90]:
grouped_sales = df.groupby('Region')['sales'].sum()
grouped_sales

Region
East      110
North     700
South    2100
West      670
Name: sales, dtype: int64

---------------------------------------------------------------------------------------------------------------------------------------------------------

4. Fill missing values in the 'Revenue' column.


In [93]:
df['Revenue'].fillna(0)

0     500.0
1    1200.0
2     800.0
3     450.0
4       0.0
5     600.0
6     670.0
Name: Revenue, dtype: float64

---------------------------------------------------------------------------------------------------------------------------------------------------------

5. Extract 'Year', 'Month', and 'Day of Week' from the 'Order Date' column

In [96]:
df['order_date'] = pd.to_datetime(df['order_date'])
# extract year
df['year'] = df['order_date'].dt.year
df['year']

0    2023
1    2023
2    2023
3    2023
4    2023
5    2023
6    2023
Name: year, dtype: int32

In [97]:
# extract month
df['month'] = df['order_date'].dt.month
df['month']

0    1
1    1
2    2
3    2
4    3
5    3
6    3
Name: month, dtype: int32

In [98]:
# extract day of week
df['day'] = df['order_date'].dt.day
df['day']

0    15
1    18
2    10
3    20
4     5
5    15
6    25
Name: day, dtype: int32

---------------------------------------------------------------------------------------------------------------------------------------------------------

6. Create a new column 'Performance': - 'High' if Sales > 1000 - 'Medium' if 500 <= Sales <=
1000 - 'Low' if Sales < 500

In [101]:
df['Performance'] = df['sales'].apply(lambda x:'High' if x > 1000 else ('Medium' if x >=500 else 'Low'))
df['Performance']

0       Low
1      High
2       Low
3       Low
4    Medium
5       Low
6    Medium
Name: Performance, dtype: object

---------------------------------------------------------------------------------------------------------------------------------------------------------

7. Return the 'Product Name' and 'Sales' where Category is 'Furniture' and Sales > 400.


In [104]:
df.loc[(df['category'] == 'Furniture') & (df['sales'] > 400),['Product_name','sales']]

Product_name  sales
3        table    450
6         sofa    670

---------------------------------------------------------------------------------------------------------------------------------------------------------

8. Create 2 one dimensional arrays and generate their dot product.


In [107]:
a = np.array([1,2,3,4,5,6])
print(a)
b = np.array([11,22,33,11,2,3])
print(b)
print(np.dot(a,b))

[1 2 3 4 5 6]
[11 22 33 11  2  3]
226


---------------------------------------------------------------------------------------------------------------------------------------------------------

9. Sort by 'Order Date' and calculate a 3 order rolling average of 'Sales'.


In [110]:
# convert order date to datetime
df['order_date'] = pd.to_datetime(df['order_date'])

df.sort_values(by='order_date')

order_id order_date customer_name Region         category Product_name  \
0      1001 2023-01-15         Alice  North        Furniture        chair   
1      1002 2023-01-18           Bob  South       Technology       laptop   
2      1003 2023-02-10       Charlie   East  Office supplier        paper   
3      1004 2023-02-20         Alice  North        Furniture        table   
4      1005 2023-03-05           Eve  South       Technology      monitor   
5      1006 2023-03-15         Frank   East  Office supplier      stapler   
6      1007 2023-03-25         Grace   West        Furniture         sofa   

   sales  quantity  Revenue  year  month  day Performance  
0    250         2    500.0  2023      1   15         Low  
1   1200         1   1200.0  2023      1   18        High  
2     80        10    800.0  2023      2   10         Low  
3    450         1    450.0  2023      2   20         Low  
4    900         2      NaN  2023      3    5      Medium  
5     30        20    600.0  2023      3   15         Low  
6    670         1    670.0  2023      3   25      Medium

In [111]:
# calculate rolling average 
df['rolling_average'] = df['sales'].rolling(3).mean()
print(df)

   order_id order_date customer_name Region         category Product_name  \
0      1001 2023-01-15         Alice  North        Furniture        chair   
1      1002 2023-01-18           Bob  South       Technology       laptop   
2      1003 2023-02-10       Charlie   East  Office supplier        paper   
3      1004 2023-02-20         Alice  North        Furniture        table   
4      1005 2023-03-05           Eve  South       Technology      monitor   
5      1006 2023-03-15         Frank   East  Office supplier      stapler   
6      1007 2023-03-25         Grace   West        Furniture         sofa   

   sales  quantity  Revenue  year  month  day Performance  rolling_average  
0    250         2    500.0  2023      1   15         Low              NaN  
1   1200         1   1200.0  2023      1   18        High              NaN  
2     80        10    800.0  2023      2   10         Low       510.000000  
3    450         1    450.0  2023      2   20         Low       576.666667 

---------------------------------------------------------------------------------------------------------------------------------------------------------

10. For each 'Category', calculate: - Total Sales - Total Quantity - Average Revenue


In [114]:
category_calculation = df.groupby('category').agg({
    'sales':'sum',
    'quantity':'sum',
    'Revenue':'mean'
}).reset_index()
print(category_calculation)

          category  sales  quantity  Revenue
0        Furniture   1370         4    540.0
1  Office supplier    110        30    700.0
2       Technology   2100         3   1200.0


11. Detect outliers in the 'Quantity' column using the IQR method.


In [116]:
q1 = df['quantity'].quantile(0.25)
q1

1.0

In [117]:
q3 = df['quantity'].quantile(0.75)
q3

6.0

In [118]:
iqr = q3 - q1
iqr

5.0

In [119]:
upperlimit = q3 + 1.5 *iqr
upperlimit

13.5

In [120]:
lowerlimit = q1 - 1.5 *iqr
lowerlimit

-6.5

In [121]:
df[df['quantity']>upperlimit]

order_id order_date customer_name Region         category Product_name  \
5      1006 2023-03-15         Frank   East  Office supplier      stapler   

   sales  quantity  Revenue  year  month  day Performance  rolling_average  
5     30        20    600.0  2023      3   15         Low            460.0

In [122]:
df[df['quantity']<lowerlimit]

Empty DataFrame
Columns: [order_id, order_date, customer_name, Region, category, Product_name, sales, quantity, Revenue, year, month, day, Performance, rolling_average]
Index: []

---------------------------------------------------------------------------------------------------------------------------------------------------------

12. Use a custom function with apply() to categorize 'Quantity' as: 'Bulk' if Quantity >= 10,
'Single' if Quantity == 1, else 'Few'


In [125]:
def categorized_qty(qty):
    if qty >= 10:
        print('Bulk')
    elif qty == 1:
        print('Single')
    else:
        print('Few')
df['category'] = df['quantity'].apply(categorized_qty)
df['category']

Few
Single
Bulk
Single
Few
Bulk
Single


0    None
1    None
2    None
3    None
4    None
5    None
6    None
Name: category, dtype: object

---------------------------------------------------------------------------------------------------------------------------------------------------------
13. Find all rows where 'Customer Name' starts with the letter 'A' or 'C'.

In [127]:
df[df['customer_name'].str.startswith('A')]

order_id order_date customer_name Region category Product_name  sales  \
0      1001 2023-01-15         Alice  North     None        chair    250   
3      1004 2023-02-20         Alice  North     None        table    450   

   quantity  Revenue  year  month  day Performance  rolling_average  
0         2    500.0  2023      1   15         Low              NaN  
3         1    450.0  2023      2   20         Low       576.666667

In [128]:
df[df['customer_name'].str.startswith('C')]

order_id order_date customer_name Region category Product_name  sales  \
2      1003 2023-02-10       Charlie   East     None        paper     80   

   quantity  Revenue  year  month  day Performance  rolling_average  
2        10    800.0  2023      2   10         Low            510.0